### PR TITLE
Use proxy-based serialization for modules

### PR DESCRIPTION
**Pull request by retronym**
This avoids serializing any state of serializable singletons which
would be discarded anyway after deserializing (SI-10412).

Use ClassValue to cache instances, which is significantly faster for
subsequent deserialization of a given module. ClassValue entries only
weakly reference the key and value so the class can be unloaded.

Wrap use of reflection in AccessControler to play nicely under
SecurityManager. A security manager will only walk up the stack
to this point to check permissions, so if users grant scala-library
permissions in the policy, the call will be allowed regardless of
the code higher in the stack that has triggered deserialization.

(Rebase and extension of #6877)

Fixes https://github.com/scala/bug/issues/10412

### CODE REVIEW

**Review by lrytz — APPROVED** [review]
Nice!

**Comment by mkeskells** [comment]
Should this be marked [nomerge] as we cant apply this to 2.12 because of serialisation binary compat
Not sure of these special tags though

**Comment by mkeskells** [comment]
As we are asserting that this work well with a SecurityManager, with minimal permissions (i.e. Scala.* has reflect access), there should be a test for that

File: src/compiler/scala/tools/nsc/typechecker/SyntheticMethods.scala [hunk 1]
```diff
@@ -326,14 +326,15 @@ trait SyntheticMethods extends ast.TreeDSL {
 
     /* If you serialize a singleton and then deserialize it twice,
      * you will have two instances of your singleton unless you implement
-     * readResolve.  Here it is implemented for all objects which have
-     * no implementation and which are marked serializable (which is true
-     * for all case objects.)
+     * readResolve.  Here we use a serialization proxy that prevents
```
Comment (szeiger) [inline src/compiler/scala/tools/nsc/typechecker/SyntheticMethods.scala]: I originally wrote this confusing comment. I think @mkeskells' version in https://github.com/scala/scala/pull/6877/files#r199624245 is better.

**Review by szeiger — APPROVED** [review]

**Comment by retronym** [comment]
I'll rebase and improve the comment.

**Comment by retronym** [comment]
@mkeskells 

> As we are asserting that this work well with a SecurityManager, with minimal permissions (i.e. Scala.* has reflect access), there should be a test for that

I can't really figure out how to (or maybe, muster the energy to) rig up a test for this. I _think_ it would require me to run a forked JVM with a security manager and a setting for `-Dpackage.access=com.foo.bar`, define an object in `com.foo.bar`, call `readResolve` from a class in an unprivileged protection domain.

> Should this be marked [nomerge] as we cant apply this to 2.12 because of serialisation binary compat. Not sure of these special tags though

That's only required for changes committed to 2.12.x that shouldn't/needn't be merged forward to 2.13.x.

**Comment by mkeskells** [comment]
@retronym I don't think that it is as complex as that for the test. I will pop one in a separate new PR

great to see this merged 😄

**Pull request merged**